JBTM-3670 Make the quickstarts work with JDK17

marcosgopen opened the pull request:
https://issues.redhat.com/browse/JBTM-3670

Failed to link org/jboss/arquillian/protocol/servlet/runner/ServletTestRunner to javax/servlet/http/HttpServlet

marcosgopen commented:
This error is caused by the new wildfly release: wildfly-27.0.0.Alpha5-SNAPSHOT.
I suggest to use wildfly-26.1.2.Final-SNAPSHOT (as also narayana is currently does) while we resolve the issue internally.

mmusgrov commented on scripts/hudson/quickstart.sh:
```diff
@@ -156,7 +156,49 @@ function build_apache-karaf {
+    rm -rf jboss-as # start afresh
+  fi
+
+  git clone $REPO jboss-as || fatal "git clone $REPO failed"
```
Maybe check that the clone succeeded here otherwise calling build_as is going to fail.

marcosgopen replied:
Thanks @mmusgrov for your comment. Doesn't the '||' operator guarantee that either the repo is cloned correctly or the fatal error is thrown?

tomjenkinson replied:
I think as @marcosgopen has it should be OK, we are using `II fatal` elsewhere too in https://github.com/jbosstm/quickstart/blob/master/scripts/hudson/quickstart.sh

mmusgrov commented on scripts/hudson/quickstart.sh:
```diff
@@ -156,7 +156,49 @@ function build_apache-karaf {
+  git checkout $AS_BRANCH || fatal "git checkout $AS_BRANCH failed"
+
+  # building WildFly
+  [ "$_jdk" -lt 17 ] && export MAVEN_OPTS="-XX:MaxMetaspaceSize=512m -XX:+UseConcMarkSweepGC $MAVEN_OPTS"
```
Where does `_jdk` get initialised.

marcosgopen replied:
Thanks @mmusgrov . I hope my last changes should fix that.

mmusgrov commented on scripts/hudson/quickstart.sh:
```diff
@@ -156,7 +156,49 @@ function build_apache-karaf {
+  export JBOSS_HOME=`echo  $JBOSS_HOME`
+
+  # init files under JBOSS_HOME before AS TESTS is started
+  init_jboss_home
```
Where is this function defined?

marcosgopen replied:
Thanks @mmusgrov . I hope my last commit fixes that.

tomjenkinson commented on scripts/hudson/quickstart.sh:
```diff
@@ -156,7 +180,61 @@ function build_apache-karaf {
+  sed -e 's#^\(.*</jboss-cli>\)#<connection-timeout>30000</connection-timeout>\n\1#' "$CONF" > "$CONF.tmp" && mv "$CONF.tmp" "$CONF"
+  grep 'connection-timeout' "${CONF}"
+  #Enable remote debugger
+  echo JAVA_OPTS='"$JAVA_OPTS -agentlib:jdwp=transport=dt_socket,address=8787,server=y,suspend=n"' >> "$JBOSS_HOME"/bin/standalone.conf
```
@marcosgopen FYI I think this is the reason the `xts-ssl` quickstart is failing at the point it is in your CI build number 111. There tries to be two servers running with port 8787

marcosgopen replied:
I had the same idea and I have commented those line, but no luck. It is about the wildfly quickstart that are downloaded here (https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh) I think and I am working on it.

tomjenkinson replied:
I think removing the exporting of the port 8787 could still be right. Could it be some other problem?

In more abstract terms, are two servers expected to be running at the same time? If so then having two servers configured with -agentlib:jdwp=transport=dt_socket,address=8787,server=y,suspend=n" at the same time will cause a problem. I can't remember explicitly what I saw but it was something like a bind exception. I can try to get the log message if it's helpful?

marcosgopen replied:
Yes, sorry you are right. I remember the bind exception. Thank you, I will change the debug port then.

tomjenkinson commented:
@marcosgopen I am not sure why but for some reason the build (at least in xts-ssl) is trying to pull a very old version of `wildfly-ee-with-tools`: [INFO] Downloading from central: https://repo.maven.apache.org/maven2/org/wildfly/bom/wildfly-ee-with-tools/20.0.0.Final/wildfly-ee-with-tools-20.0.0.Final.pom

It's strange because the version declared in https://github.com/wildfly/quickstart/blob/main/wsat-simple/pom.xml#L47 looks to be later.

Could some unpredictability be introduced by the patching here: https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh#L121-L149 ?

marcosgopen commented:
Thanks @tomjenkinson for your comment. Actually the "old version" you mentioned is old because it is a fixed one: https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh#L202

marcosgopen commented:
@tomjenkinson the problem is that we should use the wildfly quickstart version 27.0.0.Alpha1 but there is neither branch nor tag about that release and the main one is already with Jakarta namespace.

tomjenkinson commented:
> Thanks @tomjenkinson for your comment. Actually the "old version" you mentioned is old because it is a fixed one: https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh#L202 I am going to change it because as you said it is an old one.

Thank you for finding this - I guess matching https://github.com/jbosstm/narayana/blob/master/scripts/hudson/narayana.sh#L382 for now?

tomjenkinson commented:
Changing `-Dversion.server.bom=20.0.0.Final -Dversion.microprofile.bom=20.0.0.Final` to `-Dversion.server.bom=27.0.0.Alpha1 -Dversion.microprofile.bom=27.0.0.Alpha1` isn't enough I guess then?

tomjenkinson commented:
Would it work with https://github.com/wildfly/quickstart/commits/26.1.1.Final? I think we should also try to remove the -Dversion.server.bom and -Dversion.microprofile.bom overrides if possible

marcosgopen commented:
@tomjenkinson I tried 26.1.1.Final and 27.0.0.Alpha1 and with both I get:
"Caused by: org.eclipse.aether.resolution.ArtifactResolutionException: Could not find artifact org.wildfly.bom:wildfly-ee-with-tools:pom:26.1.1.Final" / 27.0.0.Alpha1
Only the main one (27.0.0.Alpha4) can download the artifacts but it is not working with wildfly 27.0.0.Alpha1 because of Jakarta namespace.

tomjenkinson commented:
Ah, I think the artifactId changed: https://repo.maven.apache.org/maven2/org/wildfly/bom/wildfly-jakartaee8-with-tools/26.1.1.Final/ (I don't think you will find 27.0.0.Alpha1 as I don't spot that tag in https://github.com/wildfly/boms/tags - but hopefully you could try to make it work with 26.1.1.Final?

tomjenkinson commented:
After https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh#L178 you could try to add:
`cd $WILDFLY_QUICKSTART_HOME`
`git checkout 26.1.1.Final`
`cd -`

And also remove `-Dversion.server.bom=20.0.0.Final -Dversion.microprofile.bom=20.0.0.Final` from (the current) https://github.com/jbosstm/quickstart/blob/master/XTS/ssl/xts-over-ssh.sh#L202

marcosgopen commented:
Thanks @tomjenkinson for your suggestion. It should work exactly as it worked before. What about if we clone the main wildfly, build it and make it run with the main wildfly/quickstart? It will work fine also for the future. I tested it locally and it works. WDYT?

marcosgopen commented on wildfly/commit-markable-resource/pom.xml:
```diff
@@ -137,12 +137,6 @@
             <version>${version.wildfly.logging}</version>
             <scope>provided</scope>
         </dependency>
-        <dependency>
```
Removed because of duplication

tomjenkinson commented:
Sorry, I am not certain what you are meaning when you say it should work as before because I think what we have as it stands now in `master` is not what we should aim to have (as in using 20.0.0.Final). Looking at https://github.com/jbosstm/quickstart/pull/403/commits/8da25b7861709288e26aad0ce322f2ea3d7ec062 I see you removed the version setting which is good but I am not familiar enough with JBOSS_BIN to know what commenting out did in relation to `main` sorry.

marcosgopen commented:
Sorry @tomjenkinson , I didn't explain myself well. What I mean is that I think it is better to use the wildfly quickstart main instead of using the old 26.1.1.Final . To do so we should build also the wildfly main branch ( commenting JBOSS_BIN forces wildfly main to be cloned and built). As a result this quickstart will work also in the future, even when we will move from 27.0.0.Alpha1 to main wildfly. 
If you think the quickstart version 26.1.1.Final is better for the moment I will apply the changes you suggested in your previous comment. Thanks

tomjenkinson commented:
Thank you @marcosgopen - I don't have a strong recommendation and if main works that works for me.

tomjenkinson commented on spring/narayana-spring-boot/pom.xml:
```diff
@@ -73,14 +77,19 @@
             <groupId>org.springframework.boot</groupId>
             <artifactId>spring-boot-starter-artemis</artifactId>
-        </dependency>
+        </dependency>      
```
FYI a bit of extra whitespace at the end of this line that should be able to be removed I expect

marcosgopen replied:
Thanks Tom, I will do it.

tomjenkinson commented on XTS/ssl/xts-over-ssh.sh:
```diff
@@ -28,7 +28,8 @@ SLEEP_TIME=`timeout_adjust 3 2>/dev/null || echo 3` # default sleep time is 3
 WORKSPACE=${WORKSPACE:-$PWD}
 WILDFLY_QUICKSTART_HOME=${WILDFLY_QUICKSTART_HOME:-$QUICKSTART_HOME}
-JBOSS_BIN=${JBOSS_BIN:-$JBOSS_HOME} # jboss_bin is unpacked jboss distribution
+#comment so main wildfly will be used (TODO: uncomment it as soon as narayana works with main wildfly)
```
I think you should raise a JBTM to get this line reverted to help remind us to revert this soon. You should reference the JBTM in the TODO comment.

marcosgopen replied:
Thank you Tom, I think it is a good idea. See the issue here: https://issues.redhat.com/browse/JBTM-3679

jmfinelli commented on ArjunaJTA/jee_transactional_app/pom.xml:
```diff
@@ -41,14 +41,10 @@
         <version.wildfly.logging>5.0.0.Beta5</version.wildfly.logging>
         <version.junit>4.13.1</version.junit>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on XTS/raw-xts-api-demo/pom.xml:
```diff
@@ -34,15 +34,14 @@
         <version.arquillian>1.1.11.Final</version.arquillian>
         <version.arquillian.graphene>2.0.3.Final</version.arquillian.graphene>
-        <version.arquillian.wildfly>1.1.0.Final</version.arquillian.wildfly>
+        <version.arquillian.wildfly>3.0.1.Final</version.arquillian.wildfly>
```
All properties specifying a version should go in the root pom

jmfinelli commented on XTS/wsat-jta-multi_hop/pom.xml:
```diff
@@ -42,7 +42,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on XTS/wsat-jta-multi_service/pom.xml:
```diff
@@ -42,7 +42,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on compensating-transactions/mongodb-simple/pom.xml:
```diff
@@ -56,7 +56,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on compensating-transactions/non-transactional_resource/pom.xml:
```diff
@@ -55,7 +55,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on compensating-transactions/travel-agent/pom.xml:
```diff
@@ -56,7 +56,7 @@
         <project.build.sourceEncoding>UTF-8</project.build.sourceEncoding>
         <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on jta-1_2-in-wildfly/pom.xml:
```diff
@@ -38,21 +38,20 @@
         <version.javax.enterprise>1.1</version.javax.enterprise>
         <version.junit>4.13.1</version.junit>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on rts/at/jta-service/pom.xml:
```diff
@@ -44,7 +44,7 @@
         <version.resteasy>3.0.23.Final</version.resteasy>
         <version.resteasy.api>3.0.5.Final</version.resteasy.api>
-        <version.org.wildfly.arquillian>2.1.1.Final</version.org.wildfly.arquillian>
+        <version.org.wildfly.arquillian>3.0.1.Final</version.org.wildfly.arquillian>
```
All properties specifying a version should go in the root pom

jmfinelli commented on wildfly/commit-markable-resource/pom.xml:
```diff
@@ -55,9 +53,9 @@
         <version.surefire.plugin>2.19.1</version.surefire.plugin>
         <version.arquillian>1.1.11.Final</version.arquillian>
-        <version.arquillian.wildfly>2.2.0.Final</version.arquillian.wildfly>
+        <version.arquillian.wildfly>3.0.1.Final</version.arquillian.wildfly>
```
All properties specifying a version should go in the root pom

jmfinelli commented on jta-1_2-in-wildfly/pom.xml:
```diff
@@ -38,21 +38,20 @@
         <jvm.args.debug></jvm.args.debug>
         <jvm.args.other>-server</jvm.args.other>
         <version.surefire.plugin>2.19.1</version.surefire.plugin>
+        <modular.jdk.args>--add-opens=java.base/java.util=ALL-UNNAMED</modular.jdk.args>
```
Shouldn't this go in a profile?

marcosgopen replied:
Thank you for your question @jmfinelli. Actually it should. But the only way to make it work is to add it to the parent profile, otherwise other profiles might not see other profile's properties. Unfortunately there is not way to establish an order of profiles in the same pom. WDYT if I create a parent pom and I put this quickstart in a subdirectory? So I could insert the profile in the 'just created' pom.

tomjenkinson replied:
I am inferring a bit here what is being asked for and what the response means but @marcosgopen, are you certain that if you don't have a jdk17 profile that defined the property modular.jdk.args then the arqIPv6 can't reliably pick it up? Something like that is done to get https://github.com/jbosstm/narayana/blob/master/txframework/pom.xml#L205 for `jdk17` and I am not sure but are failures being reported for JDK 17 builds not having --add-opens set?

jmfinelli replied:
What about if you leave the property `modular.jdk.args` empty here and then you define a `jdk17` profile (that gets active only when JDK >= 17) where you fill `modular.jdk.args` with `--add-opens`?

jmfinelli replied:
Basically
```
<properties>
    <!-- populated using the jdk17 profile -->
    <modular.jdk.args></modular.jdk.args>
</properties>
<profile> 
    <id>jdk17</id>
        <activation>
          <jdk>[17,)</jdk>
        </activation>
        <properties>
           <modular.jdk.args>--add-opens=java.base/java.util=ALL-UNNAMED</modular.jdk.args>
        </properties>
    </profile>
</profiles>
```

marcosgopen replied:
@jmfinelli that is what I tried and I expected it to work. But when I run "mvn clean install" arq profile is executed and the modular.jdk.args property is empty (jdk17 profile does not seem to be executed). That is why it fails and I ended up using the property.

marcosgopen replied:
I found out that if we define an activation property (i.e. !ipv6) in the arq profile the behaviour is different: maven wait all the profiles to define variables (i.e. ${modular.jdk.args}) and then execute arq profile correctly (quite odd behaviour) that means the modular.jdk.args variable is correctly defined. I think this might be a solution.

jmfinelli replied:
I did not dive very deep into this issue but I just tested your branch. If I define a `jdk17` profile as I suggested above, the quickstart does not activate the `arq` profile (which should be actived by default :confused: ). I agree, strange behaviour. Anyway, if `-Parq` is added to `mvn clean test` (or `./build.sh` etc.), everything works as it should. If you meant that this (i.e. `P-arq`) might be a workaround, I agree :-) OTOH, it would be nice to understand why `arq` does not get activated 

BTW, I also tried to delete `<modular.jdk.args></modular.jdk.args>` from the `properties` section and (as expected) the quickstart works fine. So I think we can only use the `jdk17` profile to define `modular.jdk.args`

marcosgopen replied:
I think you are right and the problem is the activeByDefault configuration for arq profile. If you agree I would replace it with an activation property which works best.

jmfinelli replied:
I think that activating an Arquillian profile by default is not very nice. As this configuration is giving us problem now (very strangly :confused: ), it could be a good occasion to change the behaviour, WDYT? I like how Narayana handles Arquillian profiles. Narayana defines a variable (ARQ) to run a specific Arquillian profile. This allows Narayana to run different profile based on the behaviour the user wants to test (for example, managed, embedded, IP6). We could develop something like that in our quickstarts, WDYT? (In case you like this idea, we could track this improvement with another JIRA)

tomjenkinson replied:
If you take off activeByDefault you can just activate it on the command line with `-Parq` (assuming the profile is called `arq`), you can activate multiple `-Parq,profile2,...`

I recently made a change to have a property `arqProfileActivated` so that I could activate arq and jdk17 profile only in the condition that it was JDK 17 too and otherwise only arq would activate. Hopefully see this in the changes to `scripts/hudson/narayana.sh` and `ArjunaJTA/jta/pom.xml` in https://github.com/jbosstm/narayana/commit/61aa3be08363f7dfe60c9897c0b99b005ab12520

I am clarifying here a bit because typically Narayana had different profiles rather than a variable to control what Arquillian profile is ran but I might be misfollowing and this might not be helpful and in which case please excuse the noise.

jmfinelli replied:
Narayana uses a variable in its [script](https://github.com/jbosstm/narayana/blob/f6b9472c6f20f72cb49b647cbccfc0e46985fd58/scripts/hudson/narayana.sh#L84) (`narayana.sh`). It seems that [this](https://github.com/jbosstm/narayana/commit/61aa3be08363f7dfe60c9897c0b99b005ab12520#diff-234fb7d762fdc80a3f29e8d8f4e0669ad47101d550eb3724420e91f64ee044deL538) instruction should have been using `-P$ARQ_PROF` :confused: .

tomjenkinson replied:
Thanks @jmfinelli - do you think that we should provide more than one arq profile in that jta one?

I think even with more than one Arquillian profile we can make something work. We can ask that the profile is specified in the `arqProfileActivated` property rather than `true`, e.g. in the `arq` profile:
```      <activation>
        <activeByDefault>false</activeByDefault>
        <property>
          <name>arqProfileActivated</name>
          <value>arq</value>
        </property>
      </activation>```

Then in the narayana.sh we can have:
`./build.sh -f ArjunaJTA/jta/pom.xml -fae -B -DarqProfileActivated=$ARQ_PROF $CODE_COVERAGE_ARGS "$@" test`

marcosgopen replied:
Thanks @tomjenkinson and @jmfinelli for your response. I can see an easy way to use the activation property and does not need to set an additional profile as maven argument. Here is the example: https://github.com/jbosstm/quickstart/pull/403/commits/bb4ba9856d10a6b833bb33b75dc3e8e75e06cfc4#diff-a78913a2bd252946c65e5393c9d22a7f0e360157ef53e33b22ca36ec2b3610b6R160-R162

marcosgopen replied:
I took the idea from other quickstart: https://github.com/jbosstm/quickstart/blob/master/XTS/wsat-jta-multi_service/pom.xml#L225

jmfinelli replied:
> do you think that we should provide more than one arq profile in that jta one

Thanks for highlighting this :-). Yeah, maybe could be an improvement and it would be in line with other module in ArjunaJTA (i.e. `cdi`).

About the activation of the `arq` profile, I would prefer to stick with `-P$ARQ_PROF` in the build script instead of defining a property (which is in line with other modules in Narayana)

tomjenkinson replied:
Re: the arquillian profiles in Narayana under ArjunaJTA I have created https://issues.redhat.com/browse/JBTM-3683

tomjenkinson replied:
I think we should err on having the profiles activated by users. That being said I think for my case I wanted to activate the jdk17 when Maven was running with JDK17 and if the Arquillian tests should be ran. Only in those two situations (specifically it is not necessary to activate the JDK17 profile even when using JDK 17 unless also using the `arq` profile).

I don't have a strong opinion that we shouldn't do that kind of thing I mention there where it's not straight forward, but I do agree that hidden triggers should be avoided. I think customizing the build based on the environment (rather than properties) seems OK and the alternative seems to be someone building with JDK17 needs to also tell our build that they are using jdk17 and that could be seen as a bit inefficient or something?

jmfinelli replied:
> I think we should err on having the profiles activated by users

Could you please share the reason why activating maven profiles in quickstarts should error?

tomjenkinson replied:
Sorry, I was meaning I was supporting generally having the profiles by users (but not exclusively). I have edited my previous comment.

jmfinelli commented on spring/narayana-spring-boot/pom.xml:
```diff
@@ -73,6 +77,11 @@
     </dependencyManagement>
 
     <dependencies>
+        <dependency>
+            <groupId>org.javassist</groupId>
+            <artifactId>javassist</artifactId>
+            <version>3.25.0-GA</version>
+        </dependency>
```
Just out of curiosity, why this dependency?

marcosgopen replied:
This is an interesting question @jmfinelli! This quickstart uses hibernate-core dependency (from spring-boot-starter-data-jpa). hibernate-core contains javassist version 3.22.0-GA that is not working fine with JDK17, so I added version 3.25.0-GA that works.
I would prefer to update the springframework dependencies ( here spring-boot-dependencies is at version 2.0.9.RELEASE) to a higher version but in latest version some classes used in this quickstart are missing (i.e. org.springframework.boot.jta.narayana.NarayanaRecoveryManagerBean). @mmusgrov Do you think these classes (i.e. NarayanaRecoveryManagerBean, DataSourceXAResourceRecoveryHelper, ...) should be included in next spring-boot releases?

marcosgopen replied:
I will add a comment in the pom to clarify this.

jmfinelli replied:
What about if you exclude `javassist` instead of adding it?

marcosgopen replied:
Unfortunately spring uses that dependency. Removing it causes a NoClassDefFoundError:
`org.springframework.beans.factory.BeanCreationException: Error creating bean with name 'entityManagerFactory' defined in class path resource [org/springframework/boot/autoconfigure/orm/jpa/HibernateJpaConfiguration.class]: Invocation of init method failed; nested exception is java.lang.NoClassDefFoundError: javassist/bytecode/ClassFile`

jmfinelli replied:
Thanks for double checking this +1

jmfinelli reviewed (changes requested):
A couple of points to address:

- In [this](https://groups.google.com/g/narayana-users/c/IbzVtbpEzEA) conversation on `narayana-users`, I think we have a choice to make. The root pom should contain:

  - either a list of properties specifying all dependencies' versions used in Narayana's quickstarts (especially for the shared dependencies)
  - or a `dependencyManagement` section specifying all versions of the dependencies and then repeating all used dependencies (without the version) where needed
  - I would prefer to implement both solutions (i.e. properties are used to express the preferred versions and the `dependencyManagement` section is used to express the version of the dependencies only once)
  - @tomjenkinson @mmusgrov WDYT?
- Please, bear in mind that not all `5.13.1.Final-SNAPSHOT` have been picked up during the review. Please, could you double check if I missed some of them?
- The first commit (i.e. `refactoring profiles ...`) should be squeezed together with the commit about JDK17
- As the JDK7 profiles will be removed soon, could you please open a new JBTM as a reminder?

Thanks for your PR @marcosgopen :-)

marcosgopen commented on ArjunaJTS/pom.xml:
```diff
@@ -22,6 +22,11 @@
+    <artifactId>narayana-quickstarts-all</artifactId>
+    <groupId>org.jboss.narayana.quickstart</groupId>
+    <version>5.13.1.Final-SNAPSHOT</version>
+  </parent>
```
Thanks @jmfinelli  for the review! the " <relativePath>../pom.xml</relativePath>" is the default for maven (and can be omitted). You have to specify it if the parent pom is not in the parent directory. Anyway if you think it would be clearer I can add it.

jmfinelli replied:
Oh, I did not know about that! Thanks! You can consider this message as noise then ;-)

tomjenkinson commented on jta-1_2-in-wildfly/pom.xml:
```diff
@@ -38,21 +38,20 @@
         <jvm.args.debug></jvm.args.debug>
         <jvm.args.other>-server</jvm.args.other>
         <version.surefire.plugin>2.19.1</version.surefire.plugin>
+        <modular.jdk.args>--add-opens=java.base/java.util=ALL-UNNAMED</modular.jdk.args>
```
This is related to --add-opens but in a different way. Given this is the quickstarts project that people can try to understand why things are done in certain ways, maybe we should try to justify a bit more why --add-opens is being used and whether it's just for testing purposes? WDYT?

marcosgopen replied:
Thanks @tomjenkinson for your comment. Yes, this --add-opens is needed because the wildfly version we are using needs that argument to work in our quickstart. I will add a comment about it.

tomjenkinson replied:
Thanks, I think a comment would be added for each use in each quickstart. If you can share why WildFly needs it it could be good I think.

marcosgopen replied:
I added the comments @tomjenkinson and you can follow the topic about this here: https://wildfly.zulipchat.com/#narrow/stream/174184-wildfly-developers/topic/InaccessibleObjectException.20.28add-opens.29

marcosgopen commented on ArjunaJTS/standalone/pom.xml:
```diff
@@ -36,8 +42,6 @@
 
   <properties>
     <version.narayana>5.13.1.Final-SNAPSHOT</version.narayana>
```
Do you mean the version at line 44, correct?

jmfinelli replied:
yes

marcosgopen commented:
> A couple of points to address:
> 
>     * In [this](https://groups.google.com/g/narayana-users/c/IbzVtbpEzEA) conversation on `narayana-users`, I think we have a choice to make. The root pom should contain:
>       
>       * either a list of properties specifying all dependencies' versions used in Narayana's quickstarts (especially for the shared dependencies)
>       * or a `dependencyManagement` section specifying all versions of the dependencies and then repeating all used dependencies (without the version) where needed
>       * I would prefer to implement both solutions (i.e. properties are used to express the preferred versions and the `dependencyManagement` section is used to express the version of the dependencies only once)
>       * @tomjenkinson @mmusgrov WDYT?
> 
>     * Please, bear in mind that not all `5.13.1.Final-SNAPSHOT` have been picked up during the review. Please, could you double check if I missed some of them?
> 
>     * The first commit (i.e. `refactoring profiles ...`) should be squeezed together with the commit about JDK17
> 
>     * As the JDK7 profiles will be removed soon, could you please open a new JBTM as a reminder?
> 
> 
> Thanks for your PR @marcosgopen :-)

Thanks @jmfinelli for your review, I hope I addressed all your points. At the moment I moved the properties in the root pom but I have not implemented the dependencyManagement because I don't think it might be user-friendly in the quickstart root pom, but I am open to discuss about it and change my mind. BTW I opened the [JBTM](https://issues.redhat.com/browse/JBTM-3685) as you suggested @jmfinelli, would you mind to have a look at it and see if the description fits what you did mean when you wrote "As the JDK7 profiles will be removed soon, could you please open a new JBTM as a reminder"? Thanks

jmfinelli commented on ArjunaJTS/interop/pom.xml:
```diff
@@ -17,19 +17,18 @@
+    <groupId>org.jboss.narayana.quickstart.jts</groupId>
+    <artifactId>narayana-quickstarts-jts</artifactId>
+    <version>5.13.1.Final-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
```
IIRC, you told me that `../pom.xml` is the default value, isn't it?

marcosgopen replied:
Thanks @jmfinelli ! Yes, it is the default. I think I copied it from other projects. I will remove that line.

jmfinelli commented on ArjunaJTS/interop/test-ejbs/pom.xml:
```diff
@@ -147,29 +129,5 @@
       </plugins>
    </build>
-
-   <profiles>
-      <profile>
-         <!-- When built in OpenShift the 'openshift' profile will be used when invoking mvn. -->
-         <!-- Use this profile for any OpenShift specific customization your app will need. -->
-         <!-- By default that is to put the resulting archive into the 'deployments' folder. -->
-         <!-- http://maven.apache.org/guides/mini/guide-building-for-different-environments.html -->
-         <id>openshift</id>
-         <build>
-            <plugins>
-               <plugin>
-                  <artifactId>maven-war-plugin</artifactId>
-                  <version>2.1.1</version>
-                  <configuration>
-                     <outputDirectory>deployments</outputDirectory>
-                     <warName>ROOT</warName>
-                     <!-- Java EE 6 doesn't require web.xml, Maven needs to catch up! -->
-                     <failOnMissingWebXml>false</failOnMissingWebXml>
-                  </configuration>
-               </plugin>
-            </plugins>
-         </build>
-      </profile>
-   </profiles>
```
Why have you removed this profile?

marcosgopen replied:
Thanks @jmfinelli! I removed it because this profile only has the maven-war-plugin (present in root pom). But I can see it has specific configuration so I am going to restore the profile.

jmfinelli commented on ArjunaJTA/jee_transactional_app/pom.xml:
```diff
@@ -90,20 +64,6 @@
-                <configuration>
-                    <!-- Java EE 6 doesn't require web.xml, Maven needs to catch up! -->
-                    <failOnMissingWebXml>false</failOnMissingWebXml>
-                    <archive>
-                        <manifestEntries>
-                            <Dependencies>org.jboss.jts</Dependencies>
-                        </manifestEntries>
-                    </archive>
```
Isn't this needed (I guess for WFLY?)

marcosgopen replied:
I tested it with all the profiles (default, arq and arqIPv6) and I found out that such a configuration is not needed in order to have a successful result. Should I keep that configuration anyway @jmfinelli ?

jmfinelli replied:
If you double check it works without...that's great! Thanks!

jmfinelli commented on ArjunaJTS/standalone/pom.xml:
```diff
@@ -19,9 +19,14 @@
+    <groupId>org.jboss.narayana.quickstart.jts</groupId>
+    <artifactId>narayana-quickstarts-jts</artifactId>
+    <version>5.13.1.Final-SNAPSHOT</version>
+    <relativePath>../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks again, I will remove it.

jmfinelli commented on ArjunaJTS/trailmap/pom.xml:
```diff
@@ -13,17 +13,18 @@
+      <groupId>org.jboss.narayana.quickstart.jts</groupId>
+      <artifactId>narayana-quickstarts-jts</artifactId>
+      <version>5.13.1.Final-SNAPSHOT</version>
+      <relativePath>../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks again, I will remove it.

jmfinelli commented on XTS/raw-xts-api-demo/test/pom.xml:
```diff
@@ -144,13 +150,15 @@
             <id>arq-wildfly-managed</id>
             <activation>
-                <activeByDefault>true</activeByDefault>
+                <property>
+                    <name>!no.arq</name>
+                </property>
```
I still think that this is not the right way to activate a profile in our quickstarts. I would prefer to require users to activate their preferred arquillian profile. Having said that, how does it work when users want to activate `arq-wildfly-remote` instead of `arq-wildfly-managed`? Should they specify `-Dno.arq -Parq-wildfly-remote`?

marcosgopen replied:
Thanks @jmfinelli for this comment!! This is my mistake! Seeing that jdk17 is not needed anymore in XTS I am going to restore the original status of profile activation. I agree that users should be able to activate their preferred arquillian profile as we discussed.

jmfinelli replied:
I think that we can do something better here (as we discussed with Tom) but if you prefer not to modify this part in this PR, I am alright with that as well ;-)

marcosgopen replied:
I agree with you Manuel, but I would prefer to create a specific issue for that because this PR. Would you please see this issue I have just created (https://issues.redhat.com/browse/JBTM-3686) and add some additional description (if needed) ?

jmfinelli commented on jta-1_2-in-wildfly/pom.xml:
```diff
@@ -145,7 +126,9 @@
             <id>arq</id>
             <activation>
-                <activeByDefault>true</activeByDefault>
+                <property>
+                    <name>!ipv6</name>
+                </property>
```
Same here (even though in this particular case, it makes more sense to connect `ipv6` to arquillian with or without IPv6

marcosgopen replied:
Thanks @jmfinelli ! This is also my mistake, I will restore the original version of the profile configuration.

jmfinelli commented on pom.xml:
```diff
@@ -140,37 +140,152 @@
+                <archive>
+                  <index>true</index>
+                </archive>
+                <failOnMissingWebXml>false</failOnMissingWebXml>
```
Wasn't this there only because to JDK6?

marcosgopen replied:
Thanks @jmfinelli for the question. Actually it is needed also with jdk17 for some quickstarts packaged as war. Do you think we should use this configuration only with specific projects that need it?

jmfinelli commented on pom.xml:
```diff
@@ -140,37 +140,152 @@
+              <artifactId>maven-war-plugin</artifactId>
+              <version>${version.war.plugin}</version>
+              <configuration>
+                <archive>
+                  <index>true</index>
+                </archive>
```
Was this share across all quickstarts?

marcosgopen replied:
Thanks @jmfinelli for the question. It was not, I will remove it if it is not needed: I make some tests.

jmfinelli commented on rts/lra-examples/lra-jwt/wildfly/lra-coordinator/pom.xml:
```diff
@@ -3,8 +3,13 @@
+        <groupId>org.jboss.narayana.quickstart.rts.lra</groupId>
+        <artifactId>examples-lra-jwt-wildfly</artifactId>
+        <version>5.13.1.Final-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks @jmfinelli, I will remove it.

jmfinelli commented on spring/narayana-spring-boot/pom.xml:
```diff
@@ -25,21 +25,18 @@
+        <groupId>org.jboss.narayana.quickstarts.spring</groupId>
+        <artifactId>spring-all</artifactId>
+        <version>5.13.1.Final-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks @jmfinelli ! I will remove it

jmfinelli commented on spring/stock-market-with-narayana-spring-boot/pom.xml:
```diff
@@ -25,57 +25,54 @@
+        <groupId>org.jboss.narayana.quickstarts.spring</groupId>
+        <artifactId>spring-all</artifactId>
+        <version>5.13.1.Final-SNAPSHOT</version>
+        <relativePath>../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks @jmfinelli ! I will remove that.

jmfinelli commented on wildfly/commit-markable-resource/pom.xml:
```diff
@@ -75,9 +54,10 @@
                 <configuration>
                     <skip>false</skip>
-                    <argLine>${surefire.argline}</argLine>
+                    <argLine>${surefire.argline} ${modular.jdk.args}</argLine>
```
Very strange you need to use ${modular.jdk.args} twice :confused:

marcosgopen replied:
Yes, this is strange. Actually now this quickstart does not need the modular.jdk.args (which is empty now) therefore I could remove them. WDYT?

jmfinelli replied:
Yes please :-)

jmfinelli commented on wildfly/commit-markable-resource/pom.xml:
```diff
@@ -38,32 +38,11 @@
-
         <server.args>-Dprojectdir=${project.build.directory}</server.args>
-        <surefire.argline></surefire.argline>
+        <surefire.argline>-Djdk.attach.allowAttachSelf=true</surefire.argline>
```
Aren't you declaring this in `arquillian.xml`? I am a bit confused with all these repetitions, are they needed?

marcosgopen replied:
Thanks @jmfinelli ! Yes, this is a repetition. I am going to restore the old line. Actually it is not taken from arquillian.xml but it is taken from this profile: https://github.com/jbosstm/quickstart/blob/master/wildfly/commit-markable-resource/pom.xml#L250-L263
As written in the comment:  " Byteman-Arquillian-related profiles for configuring use of Byteman on different platforms/JDKs  these setup is needed for Arquillian Extension Byteman which makes possible to use Byteman rules with @BMRule while Arquillian JUnit runner is in command "

jmfinelli commented on wildfly/setCheckedActionFactoryExample/pom.xml:
```diff
@@ -23,17 +23,16 @@
+        <artifactId>narayana-quickstarts-all</artifactId>
+        <groupId>org.jboss.narayana.quickstart</groupId>
+        <version>5.13.1.Final-SNAPSHOT</version>
+        <relativePath>../../pom.xml</relativePath>
```
Same here

marcosgopen replied:
Thanks @jmfinelli ! Actually in this case it is not the default therefore I would leave it like that.

jmfinelli replied:
Sorry! My copy and paste error ;-)

jmfinelli commented on pom.xml:
```diff
@@ -140,37 +140,152 @@
+        <version.resteasy.jaxrs>4.0.0.Beta5</version.resteasy.jaxrs>
+        <version.resteasy>3.0.23.Final</version.resteasy>
+        <version.shrinkwrap>1.2.6</version.shrinkwrap>
+        <version.surefire.plugin>2.19.1</version.surefire.plugin>
```
Could you please update this to the latest? In many quickstarts, surefire was already updated to 3.x

marcosgopen replied:
Thanks @jmfinelli ! I think it is a good idea. I would go with the latest one then, which is the version 3.0.0-M7

jmfinelli commented:
> Thanks @jmfinelli for your review, I hope I addressed all your points. At the moment I moved the properties in the root pom but I have not implemented the dependencyManagement because I don't think it might be user-friendly in the quickstart root pom, but I am open to discuss about it and change my mind. BTW I opened the [JBTM](https://issues.redhat.com/browse/JBTM-3685) as you suggested @jmfinelli, would you mind to have a look at it and see if the description fits what you did mean when you wrote "As the JDK7 profiles will be removed soon, could you please open a new JBTM as a reminder"? Thanks

Thanks for you PR @marcosgopen
I am not sure why `dependencyManagement` is not user-friendly (at least, if we compare `dependencyManagement` with specifying the same version over and over again). I do not have a strong preference though, maybe someone else can contribute with her/his opinion.

JBTM-3685 looks good to me :-) Thanks!

marcosgopen commented on XTS/raw-xts-api-demo/pom.xml:
```diff
@@ -80,4 +64,21 @@
         <module>ear</module>
         <module>test</module>
     </modules>
+
+    <profiles>
+<!--     jdk17 profile is defined because org.jboss.arquillian.graphene.shaded.net.sf.cglib.core.ReflectUtils needs -->
+<!--     java.lang add-opens in order to work with JDK17  -->
+<!--     see https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-7BB28E4D-99B3-4078-BDC4-FC24180CE82B -->
+     <profile>
+          <id>jdk17</id>
+          <activation>
+            <jdk>[17,)</jdk>
+          </activation>
+          <properties>
+             <modular.jdk.args>
+              --add-opens=java.base/java.lang=ALL-UNNAMED
+            </modular.jdk.args>
+          </properties>
+      </profile>
+  </profiles>
```
This profile is needed because this quickstart uses graphene that requires it. This is not visible from Jenkins because those tests are skipped there, but I found out it is required to be successful locally. This test is about graphene webdriver.

jmfinelli commented on XTS/raw-xts-api-demo/test/pom.xml:
```diff
@@ -62,6 +62,12 @@
             <type>pom</type>
             <scope>test</scope>
         </dependency>
+<!--         update htmlunit-core-js dependency version that is already included in graphene-webdriver but at version 2.13 -->
+        <dependency>
+            <groupId>net.sourceforge.htmlunit</groupId>
+            <artifactId>htmlunit-core-js</artifactId>
+            <version>2.64.0</version>
+        </dependency>
```
Have you tried to update `graphene-webdriver`?

marcosgopen replied:
Thanks @jmfinelli , I update the pom and I removed the graphene-webdriver that I found out it was not necessary. In fact, only the arquillian-drone-webdriver-depchain sub-dependency is now used.

jmfinelli commented on XTS/raw-xts-api-demo/pom.xml:
```diff
@@ -80,4 +64,21 @@
         <module>ear</module>
         <module>test</module>
     </modules>
+
+    <profiles>
+<!--     jdk17 profile is defined because org.jboss.arquillian.graphene.shaded.net.sf.cglib.core.ReflectUtils needs -->
+<!--     java.lang add-opens in order to work with JDK17  -->
+<!--     see https://docs.oracle.com/en/java/javase/16/migrate/migrating-jdk-8-later-jdk-releases.html#GUID-7BB28E4D-99B3-4078-BDC4-FC24180CE82B -->
+     <profile>
+          <id>jdk17</id>
+          <activation>
+            <jdk>[17,)</jdk>
+          </activation>
+          <properties>
+             <modular.jdk.args>
+              --add-opens=java.base/java.lang=ALL-UNNAMED
+            </modular.jdk.args>
+          </properties>
+      </profile>
+  </profiles>
```
Have you tried to update `org.jboss.arquillian.graphene`? `2.0.3.Final` was released in Jun 2014. I would try graphene version `2.5.4` and possibly `3.0.0-alpha.3`

marcosgopen replied:
Thanks @jmfinelli for your suggestion! I managed to update the quickstart as much as I could, but the dependency need to be a fixed version to work together. Nevertheless I managed to make it run without the need to have the modular jvm arguments for jdk 17.

jmfinelli commented on XTS/raw-xts-api-demo/test/pom.xml:
```diff
@@ -118,7 +124,7 @@
                     <forkedProcessTimeoutInSeconds>${surefire.forked.process.timeout}</forkedProcessTimeoutInSeconds>
                     <!-- System properties to forked surefire JVM which runs clients. -->
-                    <argLine>${jvm.args.ip.client}</argLine>
+                    <argLine>${jvm.args.ip.client} ${modular.jdk.args}</argLine>
```
is `${modular.jdk.args}` still needed?

marcosgopen replied:
Thanks @jmfinelli , I can remove it now.

jmfinelli reviewed (commented):
Look good to me. Before approving this, could you please try to update graphene and check if `${modular.jdk.args}` is still needed?

jmfinelli reviewed (approved):
LGTM! Thank you very much @marcosgopen for this PR
